src/prpy/perception/block_detector.py

```diff
@@ -0,0 +1,180 @@
+        
+            env = robot.GetEnv()
+
+            # Get the pr-ordata package path
+            import prpy.util
+            block_path = os.path.join('objects', 'block.kinbody.xml')
+            
+            detected_blocks = self.find_blocks(cloud_topic=self.point_cloud_topic)
+        
+            # Place blocks on the table
+            from prpy.util import ComputeEnabledAABB
+            with prpy.rave.Disabled(table, padding_only=True):
+                table_aabb = ComputeEnabledAABB(table)
+                z = table_aabb.pos()[2] + table_aabb.extents()[2] + table_z_offset #OFFSET SET AT TOP
+
+            for b in detected_blocks:
+
+                block = env.ReadKinBodyXMLFile(block_path)
+
+                for link in block.GetLinks():
+                    for geometry in link.GetGeometries():
+                        geometry.SetDiffuseColor(numpy.array([b.avg_color.r,b.avg_color.g,b.avg_color.b,b.avg_color.a]))
+
+                block_pose = numpy.array(quaternion_matrix([
+                        b.pose.orientation.x,
+                        b.pose.orientation.y,
+                        b.pose.orientation.z,
+                        b.pose.orientation.w]))
+                block_pose[0,3] = b.pose.position.x
+                block_pose[1,3] = b.pose.position.y
+                block_pose[2,3] = b.pose.position.z
+
+                self.listener.waitForTransform(
+                        self.detection_frame,
+                        self.destination_frame,
+                        rospy.Time(0),
+                        rospy.Duration(timeout))
+
+                frame_trans, frame_rot = self.listener.lookupTransform(
+                        self.destination_frame,
+                        self.detection_frame,
+                        rospy.Time(0))
+
+                frame_offset = numpy.matrix(quaternion_matrix(frame_rot))
+                frame_offset[0,3] = frame_trans[0]
+                frame_offset[1,3] = frame_trans[1]
+                frame_offset[2,3] = frame_trans[2]
+
+                block_in_world = numpy.array(numpy.dot(frame_offset,block_pose))
+              
+                
+                #Snap block to table
+                
+                block_in_world[2,3] = z
+
+                #To snap blocks to upright on table
+                
+                block_matrix = block_in_world[0:3,0:3]
+                ax, ay, az = euler_from_matrix(block_matrix)
+                ax = 0
+                ay = 0
+                block_in_world_corrected = euler_matrix(ax,ay,az)
+                block_in_world[0:3,0:3] = block_in_world_corrected[0:3,0:3]
+                
+
+                #Set block name - should we change?
+                block.SetTransform(block_in_world)
+                valid = False
+                while not valid:
+                    rand_name = 'block' + `int(numpy.random.randint(1,10000))`
+                    valid = env.GetKinBody(rand_name) is None
```

Comment:
`env.Add` takes an optional second argument `anonymous` that will append an integer to the end of the `KinBody`'s name to make it unique. Please replace this loop with that functionality.